### ISSUE TITLE
Questions about cross attention during performing SVA.

### ISSUE THREAD

**Issue by liujpg** [cambrian-mllm/cambrian#72]
I find that before performing cross-attention in vision_sampler_{} of SVA, you flattened the query to a shape of [B*token_len, num_heads, 1, dim_size], with that of the key and value is [B*token_len, num_heads, num_vision_towers, dim_size]. But the function you used, torch.nn.functional.scaled_dot_product_attention() requires a input of [B, ..., token_len, dim_size], where the B is batch_size. 
Could you please explain the reason for this operation? Thank you very much!
<img width="525" alt="image" src="https://github.com/user-attachments/assets/81f70928-e65a-463c-a052-245e21f3f674">
<img width="880" alt="image" src="https://github.com/user-attachments/assets/b3427bcf-0c81-44a5-b662-e64166787997">

**Comment by penghao-wu** [comment]
The cross-attention in SVA is performed at each position of the 2D grid independently, so the seq_length of the query should be 1.

**Comment by liujpg** [comment]
Thanks for answering! From my understanding, there are L*L grids joining the dot product，so that the seq_length should be L*L such as 576. Is that correct?

**Comment by penghao-wu** [comment]
Correct. The 2304 in your example is from 4 * 576 = 2304, where 4 is batch size.

**Comment by liujpg** [comment]
Oh i get it!!! When performing dot product, we just care about the attention between a specific grid(one query in your paper) and the corresponding patch of vision feature(whose size is mk * mk), so the token length is set to 1. The L*L(574) is merged to the dim of batch size because we don't care about the attention between tokens. Right?
